### PR TITLE
[web] Render PlatformViews with SLOT tags.

### PR DESCRIPTION
**Pull request by ditman**
Render PlatformViews in SLOT tags. To do so:

* Wraps most of the DOM from a flutter app in the Shadow DOM of flt-glass-pane (was https://github.com/flutter/engine/pull/25483)
* Unifies the Platform View lifecycle in a PlatformViewManager class. That class keeps track of platform view factory functions, their rendered output, and the SLOTs for the Shadow DOM. 
  * Previously a PlatformView was a single DOM node, but now it has been split in two:
    * The `slot`(which flutter is able to move around within the render tree).
    * The `content` (which is always a direct child of the `flt-glass-pane` element, and contains the rendered output of a platform view factory).
* Platform messages related to platform views are now handled by a `PlatformViewMessageHandler` class, which is configured in the engine's `platform_dispatcher.dart`. Platform Messages for Platform Views are now more easily testable.
* Unused code across the engine has been removed (everything related to the old way of handling platform views in canvaskit and html renderers).

### Fixes 

* https://github.com/flutter/flutter/issues/80524

### Testing

* Deployed the Gallery App using this version of the engine: https://dit-gallery.web.app
* All existing tests pass.
* Added new tests for the new classes added.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/dom_renderer.dart:466]: delete?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:454]: nit: use `??=`

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/platform_views/content_manager.dart:7]: delete?

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/platform_views/content_manager.dart:167]: nit: break this into multiple strings to avoid the long line

File: lib/web_ui/test/engine/platform_views/content_manager_test.dart
Comment (hterkelsen) [inline lib/web_ui/test/engine/platform_views/content_manager_test.dart:43]: test that `knowsViewId` is false after registering but before `renderContent`

File: lib/web_ui/test/engine/platform_views/content_manager_test.dart
Comment (hterkelsen) [inline lib/web_ui/test/engine/platform_views/content_manager_test.dart:102]: nit: you can use `same` from `package:test` here which I think explicitly uses `identical`

**Review by hterkelsen — APPROVED** [review]
Mostly LGTM with some nits and test fixes

File: lib/web_ui/test/engine/platform_views/content_manager_test.dart
Comment (ditman) [reply lib/web_ui/test/engine/platform_views/content_manager_test.dart:102]: Ah yes, by not passing anything I'm settling to `equals`. `identical` is what I wanted. I did the same thing for the `renderContent` test too.

File: lib/web_ui/test/engine/platform_views/content_manager_test.dart
Comment (ditman) [reply lib/web_ui/test/engine/platform_views/content_manager_test.dart:43]: Added an expectation right below the registerFactory to check that viewsId is not known, good catch.

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:167]: Done!

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:7]: Yes, gone!

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:454]: Yes, to me the if is still more legible, but `??=` is better.

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:466]: Indeed!

**Review by hterkelsen — APPROVED** [review]

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_views/content_manager.dart:36]: As discussed offline, let's try to remove this cache.

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_views/content_manager.dart:99]: As discussed offline, let's make this a top-level function, or inline into the renderers.

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:39]: _rootViews could store a tuple (root, slot) to simply the logic in _compositeWithParams.

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:165]: I'm having trouble finding a base class that doesn't make me pepper the rest of the code with casts, and that works at the same time with ShadowRoot and Document.

(Maybe the casts *are* necessary for this to support both root types!)

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:12]: Yes, much better! Done.

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:36]: This will be cached by the proposed `embedded_views.dart` tuple, instead of the content_manager (since this cache seems to only be required for the Canvaskit renderer)

Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:36]: (Slot cache removed from the content_manager.dart)

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:99]: Done as a top-level fn.

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (creativecreatorormaybenot) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:165]: @ditman What about creating a wrapper class that takes the element in its constructor and then only does the cast once. Now, you can add members for all properties you need (like an interface) and they are type safe because you handle both cases in the wrapper.

Comment (creativecreatorormaybenot) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:165]: (you can do it in a way that prevents all casts)

```suggestion
factory _Wrapper(element) {
  return _Wrapper(
    element is ShadowRoot ? element : null,
    element is Document ? element : null,
  );
}

final ShadowRoot? shadowRoot;
final Document? document;

T get memberA => shadowRoot?.memberA ?? document!.memberA;
```

Forgive the syntax, typed it on my phone.

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart]: Added ViewClipChain "tuple" to cache what canvaskit needs to composite a platform view (at least the html part of it)

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:165]: @creativecreatorormaybenot yes! I thought of something similar maybe with extension methods. But yeah, this is probably better a TODO for the next version of this :P

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:44]: /cc @hterkelsen this unifies _views, _rootViews, and _clipCount in a single class, and keeps it all together (this is something new)

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:332]: I'm not sure this new code will actually dispose of "disposed" views. In the old code we kept a list of views which were disposed, and then later we disposed them. In this code we are only disposing views which are trying to be composited into the scene but were never registered. In the old code we would delete disposed views whether or not they are being actively composited this frame.

**Review by hterkelsen — APPROVED** [review]
mostly LGTM

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:332]: Ah yes, the thing is that in the old code this was coupled with the creation/destruction of the actual content of the platform view in the DOM.

Now, slots are "lightweight" so they can be destroyed/recreated per frame without impacting performance, and the actual platform view (with the heavy iframe) will still get destroyed when the framework signals a disposal.

Can you think of some example where this code might be leaking unused views?

File: lib/web_ui/lib/src/engine/platform_views/slots.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_views/slots.dart:7]: nit: no need to capitalize the word "slot".
nit: `[viewId]`.

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_views/content_manager.dart:49]: Arrrgh, I did refactor with visual studio, but it left some unchanged. Good eye, I'll rewrite the docs.

**Comment by ditman** [comment]
Current test failures seem to be related to infra issues:

```
Infra Failure: Step('gcloud pubsub') (retcode: 1)
```

**Comment by ditman** [comment]
Blocked by https://github.com/flutter/flutter/pull/82926

Will rebase the branch after the above lands, to kick the CI again.

File: lib/web_ui/lib/src/engine/dom_renderer.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/dom_renderer.dart:165]: This ended up happening, here: https://github.com/flutter/engine/pull/26647

**Comment by abnerh69** [comment]
Is this chat still opened?  I am trying to avoid reload of iframe url using master/stable channel but can't make it works with slot like you said.  Please, help!

**Pull request merged**
